### ISSUE TITLE
Please clarify the usage of InheritedWidget

### ISSUE THREAD

**Issue by escamoteur** [flutter/flutter#15576]
Hi,

after long discussions in Gitter today I hope you can shed some light on that topic.

What is the correct way to use an InheritedWidget? So far I understood that it gives you the chance to propagate data down the Widget tree. In extreme if you put is as RootWidget it will be accessible from all Widgets in the tree on all Routes, which is fine because somehow I have to make my ViewModel/Model accessible for my Widgets without having to resort to globals or Singletons.

BUT InheritedWidget is immutable, so how can I update it? And more important how are my Stateful Widgets triggered to rebuild their subtrees?

Unfortunately the documentation is here very unclear and after discussion with a lot of people including @brianegan nobody seems really to know what the correct way of using it.

Thanks a lot for your great work! I really like Flutter coming from Xamarin!
Thomas

**Comment by brianegan** [comment]
Yes, I see it as a way to propagate data down the tree. What I find confusing:

"Inherited widgets, when referenced in this way, will cause the consumer to rebuild when the inherited widget itself changes state."

When I first read this, I thought:
  -  I could stuff some data in the `InheritedWidget` and mutate it later. 
  - When that mutation happens, it will rebuild all the Widgets

What I found:
  - In order to mutate the State of an InheritedWidget, you need to wrap it in a StatefulWidget
  - You then actually mutate the state of the StatefulWidget and pass this data down to the InheritedWidget, which hands it to all of it's children.

However, in that case, it seems to rebuild the entire tree underneath the `StatefulWidget`, not just the Widgets connected to the `InheritedWidget`. Is that correct?

**Comment by escamoteur** [comment]
It would be way easier if every Widget had a DataContext Property that is propagated down the tree unless a widget assign a new value to it that is propagated to its children.
That's the way Xamarin forms does it which makes implementing mvvm much easier and it's very fast to access the DataContext without any lookup.

**Comment by escamoteur** [comment]
Is this the right way to do it?
https://stackoverflow.com/questions/49491860/flutter-how-to-correctly-use-an-inherited-widget

**Comment by brianegan** [comment]
@escamoteur If you treat the `InheritedWidget` as a container to pass data down, you could consider storing either a `BehaviorSubject` (since you're an Rx fan) or a `Listenable` (such as a `ValueNotifier`) in the `InheritedWidget`.

If you use a `BehaviorSubject`, you could use a `StreamBuilder` to listen for changes to the data and rebuild Widgets. You could `add` Data to the Subject to update the data.

If you use a `ValueNotifier` you can use an [`AnimatedBuilder`](https://docs.flutter.io/flutter/widgets/AnimatedBuilder/AnimatedBuilder.html) to listen for data changes, and update it by setting a new value, e.g. `MyInheritedWidget.of(context).counter.value = 2` (which will notify all Widgets listening to the `ValueNotifier` that they need to rebuild)

**Comment by PandaGeek1024** [comment]
Wait, no one can fully explain the question here? when wrapping the inheritedWidget in a statefulWidget, the updateShouldNotify() doesn't seem to work at all. Even if I set it to return false, it  still rebuild everything in the subtree. The most confusing widget in flutter.

**Comment by mono0926** [comment]
@PandaGeek1024 

https://medium.com/flutter-jp/inherited-widget-37495200d965 should help you.
The article is written in Japanese, but you might read it by using Google Translate etc.

**Comment by darshankawar** [comment]
Hi @escamoteur,
Can you please confirm if Is this still an issue ? You may refer [this](https://api.flutter.dev/flutter/widgets/InheritedWidget-class.html) for more details on Inherited widget.

**Comment by escamoteur** [comment]
Thank, the video is great, but it does not explain how changes of an InheritedWidget triggers rebuilds

**Comment by larssn** [comment]
_I assume everyone has figured this out already, but I'll still explain how it works in case anyone else is confused._

**It's confusingly simple™**

Think of an InheritedWidget as an immutable data container. It is not the responsibility of the InheritedWidget to update it's data. That responsibility falls on the parent. The parent could be anything, as long as it builds a new InheritedWidget instance with the new data. A StreamBuilder putting data from a stream into a new instance of said InheritedWidget is a good candidate.

Now how do I get the data into a widget that depends on it? By using `context.dependOnInheritedWidgetOfExactType<SomeInheritedWidget>` in the widget that depends on said data. This will allow that child widget (no matter how far down the tree it is), to get notified whenever the InheritedWidget is replaced in the widget-tree. Think of `dependOnInheritedWidgetOfExactType<SomeInheritedWidget>` as a subscription to `SomeInheritedWidget` that persist until said child widget is disposed. Kinda like a simple form of stream!

Read the above a few times if you don't get it; it's deceptively simple.

Then read this a few times: https://api.flutter.dev/flutter/widgets/InheritedWidget-class.html and watch the video until you get it.

**Comment by dominikh** [comment]
@larssn I'm afraid that doesn't address the original question.

> This will allow that child widget (no matter how far down the tree it is), to get notified (read: rebuilt) whenever the InheritedWidget is replaced in the widget-tree

If the inherited widget is replaced, then that forces a rebuild of the subtree, anyway, explicit dependency or not. That is true for any widget. A widget being rebuilt/replaced also rebuilds its subtree. Yes, InheritedWidget allows accessing it efficiently (that is, not O(n) in the depth of the tree) from anywhere in the subtree, but that's not the contentious part. The question rather is: why does InheritedWidget track its dependencies, and when, if ever, does updateShouldNotify actually work? Or, in other words: under what circumstances does InheritedWidget notify its dependents, but does not cause a rebuild of the entire subtree?

To answer that question, as far as I understand it: it only matters when you cache widgets, either manually or using const constructors. In that case, changing the inherited widget will rebuild the tree, but you'll get identical widgets for (parts of) the subtree, which will cause Flutter to stop rebuilding and instead reuse the existing widgets. In the ideal case, the child of an InheritedWidget is const, requiring none of the tree, except for actual dependents, to be rebuilt. Without the explicit tracking of dependents, such const widgets would not update.

Note that a widget being const doesn't mean that its build function can't depend on external data (such as an InheritedWidget), only that the description of how it does so is constant.
